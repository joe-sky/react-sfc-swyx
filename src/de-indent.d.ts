declare module 'de-indent' {
  function deindent(k: string): string
  export = deindent
}
