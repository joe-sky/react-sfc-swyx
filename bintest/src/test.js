import React from 'react';export default () => <div>sldkj</div>