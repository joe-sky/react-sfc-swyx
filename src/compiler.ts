import { walk } from "estree-walker";
const { Parser } = require("acorn");
import MagicString from "magic-string";
import {
  ImportDeclaration,
  TemplateLiteral,
  ExportNamedDeclaration,
  ArrowFunctionExpression,
  ExportDefaultDeclaration,
  FunctionDeclaration,
  ReturnStatement,
  VariableDeclaration,
  VariableDeclarator,
  AssignmentExpression,
  UpdateExpression,
  Identifier,
  Literal,
  MemberExpression,
  Expression,
} from "estree";

const hooks = [
  "useState",
  "useEffect",
  "useContext",
  "useReducer",
  "useCallback",
  "useMemo",
  "useRef",
  "useImperativeHandle",
  "useLayoutEffect",
  "useDebugValue",
];

interface ASTNode {
  start: number;
  end: number;
  type: string;
}

interface JSXOpeningElement extends ASTNode {
  type: "JSXOpeningElement";
  attributes: JSXAttribute[];
  name: JSXIdentifier;
  selfClosing: boolean;
}
interface JSXClosingElement extends ASTNode {
  type: "JSXClosingElement";
  name: JSXIdentifier;
}
interface JSXElement extends ASTNode {
  type: "JSXElement";
  attributes: JSXAttribute[];
  openingElement: JSXOpeningElement;
  closingElement: JSXClosingElement;
  children: (JSXElement | JSXText)[];
}
interface JSXText extends ASTNode {
  type: "JSXText";
  value: string;
  raw: string;
}
interface JSXAttribute extends ASTNode {
  type: "JSXAttribute";
  name: JSXNamespacedName | JSXIdentifier;
  value: JSXExpressionContainer | Literal;
}
interface JSXIdentifier extends ASTNode {
  type: "JSXIdentifier";
  name: string;
}
interface JSXNamespacedName extends ASTNode {
  type: "JSXNamespacedName";
  namespace: JSXIdentifier;
  name: JSXIdentifier;
}
interface JSXExpressionContainer extends ASTNode {
  type: "JSXExpressionContainer";
  expression: Expression;
}

type CompilerOptions = {
  useStateWithLabel?: boolean;
  componentDisplayName?: string;
};

export function Compiler(args: {
  code: string;
  parser?: any; // TODO: parser type
  options?: CompilerOptions;
}) {
  // console.log({acorn})
  function defaultacorn(x: string) {
    const MyParser = Parser.extend(require("acorn-jsx")());
    return MyParser.parse(x, {
      sourceType: "module",
    });
  }
  const parser = args.parser || defaultacorn;
  const ast: ASTNode = parser(args.code);
  let ms = new MagicString(args.code);

  const isProduction = false; // TODO: figure out how to get this from rollup
  // undocumented option - tbd if we actually want to let users configure
  // TODO: can make it dev-only, or maybe also useful in prod?
  const options = args.options || ({} as CompilerOptions);
  const componentDisplayName = options.componentDisplayName;
  const userWantsUSWL = options.useStateWithLabel || !isProduction;

  let STYLECONTENT, STYLEDECLARATION: any; //  typescript is stupid about assignment inside an if, thinks this is a `never` or a `undefined`
  let STYLESTATICCSS: any; // just stubbing this out

  /** refactoring to exportDefaultNode.. remind me to delete */
  // let lastChildofDefault: any; //  typescript is stupid about assignment inside an if, thinks this is a `never` or a `undefined`
  // let pos_HeadOfDefault: number;

  let exportDefaultNode: {
    lastChildofDefault?: any;
    pos_HeadOfDefault?: number;
    node?: ExportDefaultDeclaration & ASTNode;
  } = {};
  let stateMap = new Map();
  let assignmentsMap = new Map();
  let bindValuesMap = new Map();
  let isReactImported = false;
  let hasExportDefault = false; // ensure there is a default export
  walk(ast, {
    enter(node, parent, prop, index) {
      if (node.type === "ImportDeclaration") {
        let _node = node as ImportDeclaration;
        if (_node.source.value === "react") isReactImported = true;
        // TODO: check that the name React is actually defined!
        // most people will not run into this, but you could be nasty
      }
      if (node.type === "CallExpression") {
        let _node = node as any; // CallExpression would be nice but SimpleLiteral is annoying
        if (hooks.some((hook) => _node.callee.name === hook)) {
          ms.prependLeft(_node.callee.start, "React.");
        }
      }

      if (node.type === "ExportNamedDeclaration") {
        let _node = (node as ExportNamedDeclaration).declaration;
        if (
          _node &&
          _node.type === "VariableDeclaration" &&
          _node?.declarations[0].id.type === "Identifier"
        ) {
          if (_node?.declarations[0].id.name === "STYLE") {
            let loc = _node.declarations[0].init as (
              | TemplateLiteral
              | ArrowFunctionExpression
            ) &
              ASTNode; // klooge as i believe TemplateLiteral type is incomplete
            // TODO: check return type of ArrowFunctionExpression as well

            /**
             *
             * MVP of css static export feature
             *
             */
            if (loc.type === "TemplateLiteral" && loc.quasis.length === 1) {
              STYLESTATICCSS = loc.quasis[0].value.raw;
              // TODO: namespace to scope to component
              // TODO: take care of css nesting
            }
            /**
             *
             * end MVP of css static export feature
             *
             */

            if (loc) {
              STYLEDECLARATION = node as VariableDeclaration & ASTNode;
              STYLECONTENT = ms.slice(loc.start, loc.end);
            }
            // TODO - consider whether to handle Literal (just string)
            // TODO - there are bunch of other expression types we may someday need to support (unlikely)
          }
        }
      }
      if (node.type === "ExportDefaultDeclaration") {
        let _node = (node as ExportDefaultDeclaration).declaration as
          | FunctionDeclaration
          | ArrowFunctionExpression; // TODO: consider other expressions
        exportDefaultNode.node = node as ExportDefaultDeclaration & ASTNode;
        if (_node.body.type === "BlockStatement") {
          let RSArg = (_node.body.body.find(
            (x) => x.type === "ReturnStatement"
          ) as ReturnStatement)?.argument as any; // TODO: this type is  missing JSXElement
          if (RSArg.type === "JSXElement") {
            exportDefaultNode.pos_HeadOfDefault = (_node.body as any).start + 1;
            exportDefaultNode.lastChildofDefault = RSArg.children.slice(-1)[0];
            hasExportDefault = true;
            // use start and end
          } else {
            throw new Error("not returning JSX in export default function"); // TODO: fix this?
          }
        }
        if (!hasExportDefault) {
          throw new Error(
            "a React SFC must export default a function component. None was detected. Pls file an issue if this is a mistake."
          );
        }
      }
      // usestate
      if (node.type === "VariableDeclaration") {
        let dec = (node as VariableDeclaration)
          .declarations[0] as VariableDeclarator & { init: ASTNode };
        if (dec.id.type === "Identifier" && dec.id.name.startsWith("_")) {
          stateMap.set(dec.id.name, {
            node, // for replacement
            value: ms.slice(dec.init.start, dec.init.end), // for use in templating
          });
        }
      }

      // SETSTATE
      if (node.type === "AssignmentExpression") {
        // todo: maybe only read assignmentexpressions if the LHS is in the stateMap
        let LHS = (node as AssignmentExpression).left;
        if (LHS.type === "Identifier" && LHS.name.startsWith("_")) {
          assignmentsMap.set(LHS.name, { node });
        }
      }
      if (node.type === "UpdateExpression") {
        // todo: maybe only read assignmentexpressions if the LHS is in the stateMap
        let ID = (node as UpdateExpression).argument as Identifier;
        if (ID.name.startsWith("_")) {
          assignmentsMap.set(ID.name, { node });
        }
      }

      // BINDING
      if (node.type === "JSXAttribute") {
        // // bind:value syntax - we may want to use this in future?
        // let _node = node as JSXAttribute;
        // if (
        //   _node.name.type === "JSXNamespacedName" &&
        //   _node.name.namespace.name === "bind"
        // ) {
        //   let RHSobject, RHSname;
        //   // TODO: in future - support RHS which is just a Literal? MAAAYBE, maybe not
        //   if (_node.value.type === "JSXExpressionContainer") {
        //     if (_node.value.expression.type === "Identifier") {
        //       // RHS is just an identifier
        //       RHSname = _node.value.expression.name;
        //     } else if (_node.value.expression.type === "MemberExpression") {
        //       // RHS is an object access
        //       let exp = _node.value.expression as MemberExpression & ASTNode;
        //       RHSobject = {
        //         objectName:
        //           (exp.object as Identifier).name ||
        //           ((exp.object as MemberExpression).object as Identifier).name, // either its an identifier '$foo.bar` or a memberexpression `$foo.bar.baz`
        //         fullAccessName: ms.slice(exp.start, exp.end),
        //       };
        //     } else {
        //       throw new Error(
        //         "warning - unrecognized RHS expression type in binding: " +
        //           _node.value.expression.type +
        //           ". We will probably do this wrong, pls report this along with your code"
        //       );
        //     }
        //   }

        //   bindValuesMap.set(
        //     node, // to replace
        //     {
        //       // LHSname: _node.name.name.name.slice(1), // only tested to work for 'value'. remove the leading $
        //       LHSname: _node.name.name.name, // only tested to work for 'value'. remove the leading $
        //       RHSname,
        //       RHSobject,
        //     }
        //   );
        // }

        // $ prefix syntax
        let _node = node as JSXAttribute;
        if (
          _node.name.type === "JSXIdentifier" &&
          _node.name.name.startsWith("$")
        ) {
          let RHSobject, RHSname;
          // TODO: in future - support RHS which is just a Literal? MAAAYBE, maybe not
          if (_node.value.type === "JSXExpressionContainer") {
            if (_node.value.expression.type === "Identifier") {
              // RHS is just an identifier
              RHSname = _node.value.expression.name;
            } else if (_node.value.expression.type === "MemberExpression") {
              // RHS is an object access
              let exp = _node.value.expression as MemberExpression & ASTNode;
              RHSobject = {
                objectName:
                  (exp.object as Identifier).name ||
                  ((exp.object as MemberExpression).object as Identifier).name, // either its an identifier '$foo.bar` or a memberexpression `$foo.bar.baz`
                fullAccessName: ms.slice(exp.start, exp.end),
              };
            } else {
              throw new Error(
                "warning - unrecognized RHS expression type in binding: " +
                  _node.value.expression.type +
                  ". We will probably do this wrong, pls report this along with your code"
              );
            }
          }

          bindValuesMap.set(
            node, // to replace
            {
              LHSname: _node.name.name.slice(1), // only tested to work for 'value'. remove the leading $
              RHSname,
              RHSobject,
            }
          );
        }
      }
    },
    // leave(node) {
    //   // if (node.scope) scope = scope.parent;
    // }
  });

  // validations
  if (exportDefaultNode.pos_HeadOfDefault === undefined) return;

  /* 
  
  // process it!
  
  */
  if (!isReactImported) ms.prepend(`import React from 'react';`);
  // remove STYLE and insert style jsx
  if (STYLEDECLARATION && STYLECONTENT) {
    ms.remove(STYLEDECLARATION.start, STYLEDECLARATION.end);
    if (exportDefaultNode.lastChildofDefault)
      ms.appendRight(
        exportDefaultNode.lastChildofDefault.end,
        `<style jsx>{${STYLECONTENT}}</style>`
      );
  }

  // useState
  if (stateMap.size) {
    // for each state hook
    stateMap.forEach(({ node, value }, key) => {
      ms.remove(node.start, node.end);
      let newStr;
      if (userWantsUSWL) {
        // should be 'let' bc we want to mutate it
        newStr = `\nlet [${key}, set${key}] = use${key}_State(${value})`;
        // i would like to use only one instance, of useStateWithLabel
        // https://stackoverflow.com/questions/57659640/is-there-any-way-to-see-names-of-fields-in-react-multiple-state-with-react-dev
        // but currently devtools uses the NAME OF THE HOOK for state hooks
        // rather than useDebugValue. so we do a simple alias of the hook
        ms.append(`
function use${key}_State(v) {
  const x = React.useState(v);
  ${
    isProduction
      ? "return x;"
      : `React.useDebugValue('${key}: ' + x[0]); return x;`
  }
}`);
      } else {
        // just plain useState
        // should be 'let' bc we want to mutate it
        newStr = `\nlet [${key}, set${key}] = React.useState(${value})`;
      }
      ms.appendRight(exportDefaultNode.pos_HeadOfDefault!, newStr);
    });
  }

  // setState
  if (assignmentsMap.size) {
    assignmentsMap.forEach(({ node }, key) => {
      // strategy: use comma separator to turn
      // $count = $count + 1
      // into
      // ($count = $count + 1, set$count($count))
      ms.prependLeft(node.start, "(");
      ms.appendRight(node.end, `, set${key}(${key}))`);
    });
  }

  // binding
  if (bindValuesMap.size) {
    bindValuesMap.forEach(({ LHSname, RHSname, RHSobject }, node) => {
      if (RHSobject) {
        // create new object, mutate new object, THEN set it
        // must be new object or react doesnt rerender
        ms.overwrite(
          node.start,
          node.end,
          `${LHSname}={${RHSobject.fullAccessName}} 
        onChange={e => {
          let temp = Object.assign({}, ${RHSobject.objectName});
          temp${RHSobject.fullAccessName.slice(
            RHSobject.objectName.length
          )} = e.target.${LHSname};
          set${RHSobject.objectName}(temp);
        }}`
        );
      } else if (RHSname) {
        ms.overwrite(
          node.start,
          node.end,
          `${LHSname}={${RHSname}} onChange={e => set${RHSname}(e.target.${LHSname})}`
        );
      } else {
        throw new Error("we should not get here. pls repurt this binding bug");
      }
    });
  }

  // adjust for displayName
  if (options.componentDisplayName && exportDefaultNode.node) {
    const componentDeclarationNode: any = exportDefaultNode.node.declaration;
    ms.overwrite(
      exportDefaultNode.node.start,
      exportDefaultNode.node.end,
      `const $$Component = ${ms.slice(
        componentDeclarationNode.start,
        componentDeclarationNode.end
      )};
$$Component.displayName = "${options.componentDisplayName}"
export default $$Component
      `
    );
  }

  let code = ms.toString();
  return {
    js: {
      code,
      map: null, // todo
    },
    css: {
      // TODO
      // TODO
      // TODO
      // TODO
      // TODO
      // TODO
      // TODO
      // static css
      // THIS WHOLE THING IS A MASSIVE TODO
      code: STYLESTATICCSS,
      map: null,
      // TODO
      // TODO
      // TODO
      // TODO
      // TODO
      // TODO
      // TODO
      // TODO
    },
  };
}
