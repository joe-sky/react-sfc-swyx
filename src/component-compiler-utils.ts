import { SourceMapGenerator } from 'source-map'
import { SFCDescriptor, ParseOptions, parseHTMLOptions } from './types'

const hash = require('hash-sum')
let cache = require('lru-cache')
cache = new cache(100)

const splitRE = /\r?\n/g
const emptyRE = /^(?:\/\/)?\s*$/

export function parse(options: ParseOptions): SFCDescriptor {
  const {
    source,
    filename = '',
    compiler,
    compilerParseOptions = { pad: 'line' } as parseHTMLOptions,
    sourceRoot = '',
    needMap = true
  } = options
  const cacheKey = hash(filename + source)

  let output: SFCDescriptor = cache.get(cacheKey)
  if (output) return output
  output = compiler.parseComponent(source, compilerParseOptions)
  if (needMap) {
    if (output.script && !output.script.src) {
      output.script.map = generateSourceMap(
        filename,
        source,
        output.script.content,
        sourceRoot,
        compilerParseOptions.pad
      )
    }
    if (output.styles) {
      output.styles.forEach(style => {
        if (!style.src) {
          style.map = generateSourceMap(filename, source, style.content, sourceRoot, compilerParseOptions.pad)
        }
      })
    }
  }
  cache.set(cacheKey, output)
  return output
}

function generateSourceMap(
  filename: string,
  source: string,
  generated: string,
  sourceRoot: string,
  pad?: 'line' | 'space'
) {
  // ): RawSourceMap {
  const map = new SourceMapGenerator({
    file: filename.replace(/\\/g, '/'),
    sourceRoot: sourceRoot.replace(/\\/g, '/')
  })
  let offset = 0
  if (!pad) {
    offset =
      source
        .split(generated)
        .shift()!
        .split(splitRE).length - 1
  }
  map.setSourceContent(filename, source)
  generated.split(splitRE).forEach((line, index) => {
    if (!emptyRE.test(line)) {
      map.addMapping({
        source: filename,
        original: {
          line: index + 1 + offset,
          column: 0
        },
        generated: {
          line: index + 1,
          column: 0
        }
      })
    }
  })
  return JSON.parse(map.toString())
}
