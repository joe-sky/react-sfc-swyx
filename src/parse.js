import deindent from 'de-indent'
import { parseHTML } from './html-parser'

const splitRE = /\r?\n/g
const replaceRE = /./g
const tagAliasLookup = tag =>
  ({
    script: 'script',
    style: 'style'
  }[tag] || null)

/**
 * Parse a single-file component file into an SFC Descriptor Object.
 */
// export function parseComponent(content: string, options?: Object = {}): SFCDescriptor {
export function parseComponent(content, options = {}) {
  // const sfc: SFCDescriptor = {
  const sfc = {
    script: null,
    styles: [],
    customBlocks: [],
    errors: []
  }
  let depth = 0
  // let currentBlock: ?SFCBlock = null
  let currentBlock = null

  let warn = msg => {
    sfc.errors.push(msg)
  }

  if (process.env.NODE_ENV !== 'production' && options.outputSourceRange) {
    warn = (msg, range) => {
      // const data: WarningMessage = { msg }
      const data = { msg }
      if (range.start != null) {
        data.start = range.start
      }
      if (range.end != null) {
        data.end = range.end
      }
      sfc.errors.push(data)
    }
  }

  // function start(tag: string, attrs: Array<ASTAttr>, unary: boolean, start: number, end: number) {
  function start(tag, attrs, unary, start, end) {
    if (depth === 0) {
      currentBlock = {
        type: tag,
        content: '',
        start: end,
        attrs: attrs.reduce((cumulated, { name, value }) => {
          cumulated[name] = value || true
          return cumulated
        }, {})
      }
      if (tag === 'style') {
        checkAttrs(currentBlock, attrs)
        sfc.styles.push(currentBlock)
      } else if (tag === 'script') {
        checkAttrs(currentBlock, attrs)
        sfc.script = currentBlock
      } else {
        // custom blocks
        sfc.customBlocks.push(currentBlock)
      }
    }
    if (!unary) {
      depth++
    }
  }

  // function checkAttrs(block: SFCBlock, attrs: Array<ASTAttr>) {
  function checkAttrs(block, attrs) {
    for (let i = 0; i < attrs.length; i++) {
      const attr = attrs[i]
      if (attr.name === 'lang') {
        block.lang = attr.value
      }
      if (attr.name === 'scoped') {
        block.scoped = true
      }
      if (attr.name === 'module') {
        block.module = attr.value || true
      }
      if (attr.name === 'src') {
        block.src = attr.value
      }
    }
  }

  // function end(tag: string, start: number) {
  function end(tag, start) {
    if (depth === 1 && currentBlock) {
      currentBlock.end = start
      let text = content.slice(currentBlock.start, currentBlock.end)
      if (options.deindent !== false) {
        text = deindent(text)
      }
      // pad content so that linters and pre-processors can output correct
      // line numbers in errors and warnings
      // if (currentBlock.type !== 'template' && options.pad) {
      if (options.pad) {
        text = padContent(currentBlock, options.pad) + text
      }
      currentBlock.content = text
      currentBlock = null
    }
    depth--
  }

  // function padContent(block: SFCBlock, pad: true | 'line' | 'space') {
  function padContent(block, pad) {
    if (pad === 'space') {
      return content.slice(0, block.start).replace(replaceRE, ' ')
    } else {
      const offset = content.slice(0, block.start).split(splitRE).length
      const padChar = block.type === 'script' && !block.lang ? '//\n' : '\n'
      return Array(offset).join(padChar)
    }
  }

  parseHTML(content, {
    warn,
    start,
    end,
    outputSourceRange: options.outputSourceRange
  })

  return sfc
}
