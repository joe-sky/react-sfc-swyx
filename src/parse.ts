import deindent from 'de-indent'
import { parseHTML } from './html-parser'
import { SFCDescriptor, SFCBlock, parseHTMLOptions, maybenum, ASTAttr } from './types'

const splitRE = /\r?\n/g
const replaceRE = /./g
// const tagAliasLookup = (tag: string) =>
//   ({
//     script: 'script',
//     style: 'style'
//   }[tag] || null)

/**
 * Parse a single-file component file into an SFC Descriptor Object.
 */
export function parseComponent(content: string, options: parseHTMLOptions = {}): SFCDescriptor {
  const sfc = {
    script: null as SFCBlock | null,
    styles: [] as SFCBlock[],
    customBlocks: [] as SFCBlock[],
    errors: [] as WarningMessage[]
  }
  let depth = 0
  let currentBlock: SFCBlock | null = null

  let warn = (msg: string, range?: { start: maybenum; end: maybenum }) => {
    sfc.errors.push({ msg })
  }

  type WarningMessage = {
    msg: string
    start?: number
    end?: number
  }
  if (process.env.NODE_ENV !== 'production' && options.outputSourceRange) {
    warn = (msg, range) => {
      const data: WarningMessage = { msg }
      if (range && range.start != null) {
        data.start = range.start
      }
      if (range && range.end != null) {
        data.end = range.end
      }
      sfc.errors.push(data)
    }
  }

  function start(tag: string, attrs: Array<ASTAttr>, unary: boolean, start: number, end: number) {
    if (depth === 0) {
      currentBlock = {
        type: tag,
        content: '',
        start: end,
        attrs: attrs.reduce((cumulated, { name, value }) => {
          cumulated[name] = value || true
          return cumulated
        }, {})
      }
      if (tag === 'style') {
        checkAttrs(currentBlock, attrs)
        sfc.styles.push(currentBlock)
      } else if (tag === 'script') {
        checkAttrs(currentBlock, attrs)
        sfc.script = currentBlock
      } else {
        // custom blocks
        sfc.customBlocks.push(currentBlock)
      }
    }
    if (!unary) {
      depth++
    }
  }

  function checkAttrs(block: SFCBlock, attrs: Array<ASTAttr>) {
    for (let i = 0; i < attrs.length; i++) {
      const attr = attrs[i]
      if (attr.name === 'lang') {
        block.lang = attr.value
      }
      if (attr.name === 'scoped') {
        block.scoped = true
      }
      if (attr.name === 'module') {
        block.module = attr.value || true
      }
      if (attr.name === 'src') {
        block.src = attr.value
      }
    }
  }

  function end(tag: string, start: number) {
    if (depth === 1 && currentBlock) {
      currentBlock.end = start
      let text = content.slice(currentBlock.start, currentBlock.end)
      if (options.deindent !== false) {
        text = deindent(text)
      }
      // pad content so that linters and pre-processors can output correct
      // line numbers in errors and warnings
      // if (currentBlock.type !== 'template' && options.pad) {
      if (options.pad) {
        text = padContent(currentBlock, options.pad) + text
      }
      currentBlock.content = text
      currentBlock = null
    }
    depth--
  }

  function padContent(block: SFCBlock, pad: true | 'line' | 'space') {
    if (pad === 'space') {
      return content.slice(0, block.start).replace(replaceRE, ' ')
    } else {
      const offset = content.slice(0, block.start).split(splitRE).length
      const padChar = block.type === 'script' && !block.lang ? '//\n' : '\n'
      return Array(offset).join(padChar)
    }
  }

  parseHTML(content, {
    warn,
    start,
    end,
    outputSourceRange: options.outputSourceRange
  })

  return sfc
}
