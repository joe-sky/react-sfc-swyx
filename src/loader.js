import { getOptions } from 'loader-utils'
import { parseComponent } from './parse'
export default function loader(source) {
  const options = getOptions(this)
  console.log({ source: parseComponent(source) })
  source = source.replace(/\[name\]/g, options.name)
  return `export default ${JSON.stringify(source)}`
}

export const sdsd = {
  styles: [
    {
      type: 'style',
      content: '\ndiv {\n  border: 1px solid red\n}\n',
      start: 92,
      attrs: {},
      end: 131
    }
  ],
  customBlocks: [
    {
      type: 'jsx',
      content: '\nexport default function MyComponent() {\n  return <div>hi</div>\n}\n',
      start: 5,
      attrs: {},
      end: 77
    }
  ],
  errors: []
}
