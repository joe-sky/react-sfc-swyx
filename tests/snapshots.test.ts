import fs from "fs";
// import path from "path";
import { Compiler } from "../src/compiler";
fs.readdirSync("tests")
  .filter((p) => fs.lstatSync("tests/" + p).isDirectory())
  .filter(p => !p.startsWith('__snapshots__'))
  .forEach((p) => {
    // console.log({ p });
    it(`passes ${p} snapshot test`, () => {
      const src = fs.readFileSync("tests/" + p + "/src.react", "utf8");
      let options
      if (fs.existsSync(`tests/${p}/options.json`)) {
        options = JSON.parse(fs.readFileSync(`tests/${p}/options.json`, 'utf8'))
        console.log({options})
      }
      const output = Compiler({ code: src, options });
      expect(output).toMatchSnapshot();
    });
  });
