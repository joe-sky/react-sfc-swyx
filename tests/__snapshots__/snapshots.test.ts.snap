// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`passes binding-nested-obj snapshot test 1`] = `
Object {
  "css": Object {
    "code": undefined,
    "map": null,
  },
  "js": Object {
    "code": "import * as React from 'react'







export default () => {
let [_count, set_count] = use_count_State(23)
let [_text, set_text] = use_text_State({
  foo: 1,
  moo: {
    djs: 3
  }
})
  React.useEffect(() => console.log('rerendered')) // no need for React import
  return (
  <div>
    Some Text
  	<input value={_text.moo.djs} 
        onChange={e => {
          let temp = Object.assign({}, _text);
          temp.moo.djs = e.target.value;
          set_text(temp);
        }} />
    <MyButton onClick={() => (++_count, set_count(_count))}>
      Count {_count}
    </MyButton>
  <style jsx>{\`
div {
	color: \${_count > 5 ? 'blue' : 'green'};
}
\`}</style></div>
  )
}


// I can define inline Components like normal
function MyButton({onClick}) {
  return <button className=\\"Button\\" {...{onClick}}>{children}</button>
}
function use_count_State(v) {
  const x = React.useState(v);
  React.useDebugValue('_count: ' + x[0]); return x;
}
function use_text_State(v) {
  const x = React.useState(v);
  React.useDebugValue('_text: ' + x[0]); return x;
}",
    "map": null,
  },
}
`;

exports[`passes options-displayName snapshot test 1`] = `
Object {
  "css": Object {
    "code": undefined,
    "map": null,
  },
  "js": Object {
    "code": "import React from 'react';const $$Component = () => {
  return <div>div</div>
};
$$Component.displayName = \\"MySFC\\"
export default $$Component
      ",
    "map": null,
  },
}
`;

exports[`passes state-simple snapshot test 1`] = `
Object {
  "css": Object {
    "code": undefined,
    "map": null,
  },
  "js": Object {
    "code": "import React from 'react';



export default () => {
let [_count, set_count] = use_count_State(0)
  return <button onClick={() => (_count++, set_count(_count))}>Click {_count}<style jsx>{\`
    button {
      // scoped by default
      background-color: \${_count > 5 ? 'red' : 'papayawhip'};
    }
  \`}</style></button>
}
function use_count_State(v) {
  const x = React.useState(v);
  React.useDebugValue('_count: ' + x[0]); return x;
}",
    "map": null,
  },
}
`;

exports[`passes styles-static snapshot test 1`] = `
Object {
  "css": Object {
    "code": "
div {
  // scoped by default
  background-color: papayawhip;
  .Button {
    border-color: cadetblue;
  }
}
",
    "map": null,
  },
  "js": Object {
    "code": "import React from 'react';

export default ({ onClick }) => {
  React.useEffect(() => console.log(\\"rerendered\\")); // no need for React import
  return (
    <div>
      Some Text
      <MyButton {...{ onClick }}>My Call To Action</MyButton>
    <style jsx>{\`
div {
  // scoped by default
  background-color: papayawhip;
  .Button {
    border-color: cadetblue;
  }
}
\`}</style></div>
  );
};

// I can define inline Components like normal
function MyButton({onClick}) {
  return <button className=\\"Button\\" {...{onClick}}>{children}</button>
}",
    "map": null,
  },
}
`;

exports[`passes temp snapshot test 1`] = `
Object {
  "css": Object {
    "code": undefined,
    "map": null,
  },
  "js": Object {
    "code": "import * as React from \\"react\\";





export default () => {
let [_count, set_count] = use_count_State(23)
  React.useEffect(() => console.log(\\"rerendered\\")); // no need for React import
  return (
    <div>
      Some Text
      {/* <input bind:value={_text.moo.djs} /> */}
      <MyButton onClick={() => (++_count, set_count(_count))}>Count {_count}</MyButton>
    <style jsx>{\`
div {
	color: \${_count > 5 ? \\"blue\\" : \\"green\\"}
}
\`}</style></div>
  );
};

// I can define inline Components like normal
function MyButton({ onClick }) {
  return (
    <button className=\\"Button\\" {...{ onClick }}>
      {children}
    </button>
  );
}

function use_count_State(v) {
  const x = React.useState(v);
  React.useDebugValue('_count: ' + x[0]); return x;
}",
    "map": null,
  },
}
`;
